File: src/main/java/io/github/oleksivio/telegram/bot/core/controller/processor/session/SessionManager.java
package io.github.oleksivio.telegram.bot.core.controller.processor.session;

import io.github.oleksivio.telegram.bot.api.model.objects.Update;
import io.github.oleksivio.telegram.bot.api.model.result.HandlerResult;
import io.github.oleksivio.telegram.bot.core.controller.handler.Handler;

import java.util.List;
import java.util.Optional;

public abstract class SessionManager {
    private final OrderManager orderManager;
    private final Handler initialHandler;
    private final Handler cancelHandler;

    public SessionManager(OrderManager orderManager,
                          Handler initialHandler,
                          Handler cancelHandler) {
        this.orderManager = orderManager;
        this.initialHandler = initialHandler;
        this.cancelHandler = cancelHandler;
    }

    public HandlerResult init(Update update) {
        return getIdOptional(update)
                .filter(id -> initialHandler.hasSubscription(update))
                .map(id -> {

                    HandlerResult handlerResult = initialHandler.run(update);
                    if (handlerResult.hasSuccess()) {
                        orderManager.next(id);
                    }
                    return handlerResult;
                })
                .orElse(HandlerResult.pass());

    }

    protected abstract Optional<Long> getIdOptional(Update update);

    public HandlerResult receive(Update update) {

        Optional<Long> userFromIdOptional = getIdOptional(update);

        if (!userFromIdOptional.isPresent()) {
            return HandlerResult.pass();
        }

        long id = userFromIdOptional.get();

        if (cancelHandler.hasSubscription(update)) {
            orderManager.reset(id);
            return cancelHandler.run(update);
        }

        List<Handler> handlerList = orderManager.getCurrent(id);
        if (handlerList.isEmpty()) {
            orderManager.reset(id);
            return HandlerResult.pass();
        }

        for (Handler handler : handlerList) {
            if (handler.hasSubscription(update)) {
                HandlerResult handlerResult = handler.run(update);

                if (handlerResult.hasSuccess()) {
                    orderManager.next(id);
                    return handlerResult;
                }
                if (handlerResult.hasCancelSession()) {
                    orderManager.reset(id);
                    return handlerResult;
                }

                return HandlerResult.error();
            }
        }

        return HandlerResult.pass();
    }

    public boolean isActive(Update update) {
        return getIdOptional(update)
                .map(orderManager::isActive)
                .orElse(false);
    }

}

File: src/main/kotlin/io/github/oleksivio/telegram/bot/core/controller/processor/session/SessionManager.kt
package io.github.oleksivio.telegram.bot.core.controller.processor.session

import io.github.oleksivio.telegram.bot.api.model.objects.Update
import io.github.oleksivio.telegram.bot.api.model.result.HandlerResult
import io.github.oleksivio.telegram.bot.core.controller.handler.Handler

abstract class SessionManager(private val orderManager: OrderManager,
                              private val initialHandler: Handler<*>,
                              private val cancelHandler: Handler<*>) {

    fun init(update: Update): HandlerResult {
        val answerId = getIdOptional(update) ?: return HandlerResult.pass()

        if (initialHandler.hasSubscription(update)) {
            val handlerResult = initialHandler.run(update)
            if (handlerResult.hasSuccess()) {
                orderManager.next(answerId)
            }
            return handlerResult
        }

        return HandlerResult.pass()

    }

    protected abstract fun getIdOptional(update: Update): Long?

    fun receive(update: Update): HandlerResult {

        val id = getIdOptional(update) ?: return HandlerResult.pass()

        if (cancelHandler.hasSubscription(update)) {
            orderManager.reset(id)
            return cancelHandler.run(update)
        }

        val handlerList = orderManager.getOrderHandlerList(id)

        for (handler in handlerList) {
            if (handler.hasSubscription(update)) {
                val handlerResult = handler.run(update)

                if (handlerResult.hasSuccess()) {
                    orderManager.next(id)
                    return handlerResult
                }
                if (handlerResult.hasCancelSession()) {
                    orderManager.reset(id)
                    return handlerResult
                }

                return HandlerResult.error()
            }
        }


        return orderManager.getErrorHandler(id)?.run(update) ?: HandlerResult.pass()
    }

    fun isActive(update: Update): Boolean {
        val id = getIdOptional(update) ?: return false
        return orderManager.isActive(id)
    }

}